nanotune/device/device.py

```diff
@@ -1,57 +1,186 @@
+from __future__ import annotations
 import copy
 import logging
+from enum import Enum
 from typing import (
-    Any, Dict, List, Optional, Sequence, Tuple, Union, Mapping, Sequence,
+    Any, Dict, Optional, Sequence, Tuple, Union, Mapping, Sequence, List
 )
-from enum import Enum
-from collections import namedtuple
-from functools import partial
+
 import numpy as np
-from dataclasses import dataclass
+from dataclasses import asdict, dataclass
 import qcodes as qc
 from qcodes import validators as vals
 from qcodes.station import Station
 
 from nanotune.device.device_channel import DeviceChannel
-if not qc.__version__.startswith('0.27'):
-    from nanotune.device.delegate_channel_instrument import DelegateChannelInstrument as DelegateInstrument
-else:
-    from qcodes.instrument.delegate import DelegateInstrument  # type: ignore
+from qcodes.instrument.delegate import DelegateInstrument
 
+from qcodes.instrument.delegate.grouped_parameter import GroupedParameter
 
 logger = logging.getLogger(__name__)
-READOUTMETHODS = ['transport', 'sensing', 'rf']
 
 nrm_cnst_tp = Mapping[str, Tuple[float, float]]
 voltage_range_type = Dict[int, Sequence[float]]
-# TODO: Use dataclass for normalization constants
-# @dataclass
-# class NormalizationConstants:
-#     transport: Sequence[float]
-#     sensing: Sequence[float]
-#     rf: Sequence[float]
-
 
-def readout_formatter(
-    *values: Any,
-    param_names: List[str],
-    name: str,
-    **kwargs: Any,
-) -> Any:
-    return namedtuple(name, param_names)(*values, **kwargs)
 
+@dataclass
+class NormalizationConstants:
+    """Container to hold normalization constant. They are the highest and
+    lowest signals measured, corresponding to pinched-off and open signals.
+    They are typically measured by setting all gates to their lowest allowed
+    values to measure the lower bound, while the upper bound is measured with
+    all gates set to their highest allowed voltages.
 
-class Device(DelegateInstrument):
+    Attributes:
+        transport (Tuple[float, float]): constants for DC transport.
+        sensing (Tuple[float, float]): constants for charge sensing (transport
+            or other).
+        rf (Tuple[float, float]): constants for reflectometry measurements.
     """
-    device_type: str, e.g 'doubledot'
-    readout = {
-        'transport': qc.Parameter,
-        'sensing': qc.Parameter,
-        'rf': qc.Parameter,
-    }
+    transport: Tuple[float, float] = (0., 1.)
+    sensing: Tuple[float, float] = (0., 1.)
+    rf: Tuple[float, float] = (0., 1.)
 
+    def update(
+        self,
+        new_constants: Union[
+            Dict[str, Sequence[float]], NormalizationConstants],
```

Comment:
I recommend using the generic Mapping rather than Dict for typing input. That makes it possible to pass in other dict like structures (defaultdict and ordereddict for example) (you would also need to update the logic below for the is instance check )